crates/lib/src/execution_data.rs

```diff
@@ -1,58 +1,121 @@
+use enum_as_inner::EnumAsInner;
+use num::complex::{Complex32, Complex64};
 use std::collections::HashMap;
-use std::convert::TryFrom;
-use std::num::{ParseIntError, TryFromIntError};
+use std::num::ParseIntError;
 use std::str::FromStr;
 use std::time::Duration;
 
-use qcs_api_client_grpc::models::controller::ReadoutValues;
+use ndarray::prelude::*;
+
+use qcs_api_client_grpc::models::controller::{readout_values, ReadoutValues};
 use quil_rs::instruction::MemoryReference;
 
 use crate::RegisterData;
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// The result of executing an [`Executable`](crate::Executable)
 #[derive(Debug, Clone, PartialEq)]
-pub struct Qvm {
+pub struct ExecutionData {
     /// The readout data that was read from the [`Executable`](crate::Executable).
     /// Key is the name of the register, value is the data of the register after execution.
-    pub registers: HashMap<Box<str>, RegisterData>,
+    pub readout_data: ReadoutMap,
     /// The time it took to execute the program on the QPU, not including any network or queueing
     /// time. If paying for on-demand execution, this is the amount you will be billed for.
     ///
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
 
+/// An enum representing every possible readout type
+#[derive(Clone, Copy, Debug, EnumAsInner, PartialEq)]
+pub enum ReadoutValue {
```

Comment:
Should we simplify this to something like: `Integer(i32), Complex(Complex64), Real(f64)`? It would be a lot easier to deal with as a user, but less efficient.

Reply:
I'll let someone with more QC domain knowledge chime in, but from a Rust perspective I don't see a problem with having only one integer type, etc. If you expect the data to be a smaller integer, it's not much extra work to convert it while you make sure it *is* an integer, for example.

Reply:
Based on [readout.proto](https://github.com/rigetti/qcs-api-client-rust/blob/main/qcs-api-client-grpc/proto/controller/readout.proto) I don't even see a `Real` case

Reply:
Yeah, I'm probably just gonna do it unless I hear otherwise, it's a bit of a pain to have to worry about 3 integer types every time you want to `match` a readout value. Python uses 32 bit ints and 64 bit floats anyways.

Reply:
> Based on [readout.proto](https://github.com/rigetti/qcs-api-client-rust/blob/main/qcs-api-client-grpc/proto/controller/readout.proto) I don't even see a `Real` case

That bit confused me too, but Mark eventually helped me make sense of it. Real QPUs only return integers or complex numbers because those are the two only types that make sense to describe a Qubit (either a Qubit was measured into `0` or `1`, or it's superposition can be described as a complex number).

The QVM complicates things, because it'll also return parameter values back to you, so if your Quil has `DECLARE theta REAL[1]`, you'll get a "readout" register like: `{"theta": 1.0}`. That's why we have to include a "Real" as a possible readout type.

Reply:
So the three flavors of data here are:

- Raw ADC values - i16
- Unclassified, integrated - Complex64 IIRC (this is a single value collected from a sequence of the raw values)
- Classified - i1 in practice today, but we should leave room for higher values (i.e. 2)

@jselig-rigetti is on the money here: those are the only values we need to support *today*, in the file he linked.

However, here's the catch - that's not _inherently_ the case, because these are built & returned from a post-processing pipeline running within the execution service. The values defined above are the types which are returned today. However, it's not unthinkable that we'd want to return different data types in the future. So, it seems reasonable to me - if it doesn't add tech debt here - to leave room for `f64`.

> that's already returned in the biggest possible types anyways.

True - over the wire, this is helped by protobuf's [varint](https://developers.google.com/protocol-buffers/docs/encoding), but once it's unpacked in Rust, it is expanded to a larger value, so we might as well just pick that same size and go with it.

The three types that @MarquessV proposes seem wise to me.

Reply:
I simplified these to the types described above.